### PR TITLE
Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell

### PR DESCRIPTION
**Pull request by raghuramank100**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39920 Enable dynamic quantized GRU support
* **#37159 Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell**
* #38851 Refactor LSTM tests

Enable dynamic quantization for GRU, LSTMCell, RNNCell and GRUCell

Differential Revision: [D20978736](https://our.internmc.facebook.com/intern/diff/D20978736/)

### CODE REVIEW

**Comment by jamesr66a** [comment]
Can you also add deprecation warnings to `torch/jit/quantized.py` in these functions: https://github.com/pytorch/pytorch/blob/9b3911c0737d1d12650cb668dc40bb979bc81eb8/torch/jit/quantized.py#L491

Once this patch lands we should move all users over to this IMO. It's getting to be an increasing maintenance burden to keep the `jit/quantized.py` code around. Thanks!

File: aten/src/ATen/native/RNN.cpp [hunk 1]
```diff
@@ -1320,6 +1319,7 @@ bool _use_cudnn_rnn_flatten_weight() {
         std::move(packed_output.data), std::move(std::get<1>(result)));     \
   }
 
+
```
Comment (supriyar) [inline aten/src/ATen/native/RNN.cpp:1322]: nit: extra line

File: aten/src/ATen/native/RNN.cpp [hunk 2]
```diff
@@ -1368,15 +1368,15 @@ std::tuple<Tensor, Tensor> quantized_gru_data_legacy(
...
-      data,
-      batch_sizes,
-      hx,
-      std::move(params),
-      has_biases,
-      num_layers,
-      dropout_p,
-      train,
-      bidirectional);
+                            data,
```
Comment (supriyar) [inline aten/src/ATen/native/RNN.cpp:1371]: formatting seems off

File: test/quantization/test_quantize.py [hunk 3]
```diff
@@ -677,51 +678,71 @@ def checkQuantized(model):
...
+            x = torch.tensor([[100, -155],
+                             [-155, 100],
+                             [100, -155]], dtype=torch.float).unsqueeze(0).repeat(niter, 1, 1)
+
+
+            for dtype in [torch.qint8, torch.float16]:
+                if torch.backends.quantized.engine == 'qnnpack' and dtype == torch.float16:
+                    continue                            # fp16 dynamic quant is not supported for qnnpack
+
+                print(dtype)
```
Comment (supriyar) [inline test/quantization/test_quantize.py:697]: remove print statement?

File: test/quantization/test_quantize.py [hunk 4]
```diff
@@ -761,12 +782,50 @@ def test_per_channel_lstm_quantize(self):
...
+            x = torch.tensor([[100, -155],
+                             [-155, 100],
+                             [100, -155]], dtype=torch.float)
+
+
+            for dtype in [torch.qint8, torch.float16]:
+                if torch.backends.quantized.engine == 'qnnpack' and dtype == torch.float16:
+                    continue                            # fp16 dynamic quant is not supported for qnnpack
+
+                print(dtype)
```
Comment (supriyar) [inline test/quantization/test_quantize.py:807]: extra print here too

**Review by supriyar — APPROVED** [review]
lgtm, left a few comments

File: test/quantization/test_quantize.py [hunk 5]
```diff
@@ -677,51 +678,71 @@ def checkQuantized(model):
...
+            model = RNNDynamicModel(module_type).eval()
+            niter = 10
+            x = torch.tensor([[100, -155],
+                             [-155, 100],
+                             [100, -155]], dtype=torch.float).unsqueeze(0).repeat(niter, 1, 1)
+
+
+            for dtype in [torch.qint8, torch.float16]:
+                if torch.backends.quantized.engine == 'qnnpack' and dtype == torch.float16:
+                    continue                            # fp16 dynamic quant is not supported for qnnpack
```
[the remark this answers is not in the document]
Comment (raghuramank100) [reply test/quantization/test_quantize.py:695]: Not sure if it is off here?

File: test/quantization/test_quantize.py (hunk 3 above)
Comment (raghuramank100) [reply test/quantization/test_quantize.py:697]: Sure, will clean this up prior to landing

**Pull request closed without merging**